conn.go

```diff
@@ -806,46 +848,75 @@ func (cn *conn) Query(query string, args []driver.Value) (_ driver.Rows, err err
 	}
 	defer cn.errRecover(&err)
 
+	var r *rows
+
 	// Check to see if we can use the "simpleQuery" interface, which is
 	// *much* faster than going through prepare/exec
 	if len(args) == 0 {
-		return cn.simpleQuery(query)
-	}
-
-	if cn.binaryParameters {
+		r, err = cn.simpleQuery(query)
+	} else if cn.binaryParameters {
 		cn.sendBinaryModeQuery(query, args)
 
 		cn.readParseResponse()
 		cn.readBindResponse()
-		rows := &rows{cn: cn}
-		rows.colNames, rows.colFmts, rows.colTyps = cn.readPortalDescribeResponse()
+		r = &rows{cn: cn}
+		r.colNames, r.colFmts, r.colTyps = cn.readPortalDescribeResponse()
 		cn.postExecuteWorkaround()
-		return rows, nil
 	} else {
 		st := cn.prepareTo(query, "")
 		st.exec(args)
-		return &rows{
+		r = &rows{
 			cn:       cn,
 			colNames: st.colNames,
 			colTyps:  st.colTyps,
 			colFmts:  st.colFmts,
-		}, nil
+		}
+	}
+
+	if r != nil && ctx.Done() != context.Background().Done() {
```

Comment:
It looks like this is trying to avoid the overhead of an extra goroutine in the case where the context will never be canceled. It would be better to write it as:

```go
if r != nil && ctx.Done() != nil
```

Same for the other places that do this.